test/metabase/query_processor_test/cast_test.clj

```diff
@@ -561,6 +561,26 @@
                     ud (-> uncasted-value java.time.Instant/parse)]
                 (is (= ud cd))))))))))
 
+(deftest ^:parallel date-truncate-datetime
+  (mt/test-drivers (mt/normal-drivers-with-feature :expressions/date)
+    (let [mp (mt/metadata-provider)]
+      (doseq [[table fields] [[:orders [{:field :created_at}]]]
+              {:keys [field]} fields]
+        (testing (str "truncating " table "." field " to date")
+          (let [field-md (lib.metadata/field mp (mt/id table field))
+                query (-> (lib/query mp (lib.metadata/table mp (mt/id table)))
+                          (lib/with-fields [field-md])
+                          (lib/expression "DATETRUNC" (lib/date field-md))
+                          (lib/limit 100))
+                result (-> query qp/process-query)
+                cols (mt/cols result)
+                rows (mt/rows result)]
+            (is (types/field-is-type? :type/Date (last cols)))
+            (doseq [[uncasted-value casted-value] rows]
+              (let [cd (-> casted-value   java.time.Instant/parse (.atZone (java.time.ZoneId/of "UTC")) .toLocalDate)
+                    ud (-> uncasted-value java.time.Instant/parse (.atZone (java.time.ZoneId/of "UTC")) .toLocalDate)]
+                (is (= ud cd))))))))))
```

Comment:
I guess this is sanity checking that the value didn't get mangled? Would it also make sense to check that it was actually truncated? Maybe the `:type/Date` check above is sufficient...

Reply:
Good catch. I wanted to do that, but I forgot.